kythe/go/extractors/config/config_selector.go

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2018 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+// This subpart of package config handles getting a
+// kythe.proto.ExtractionConfiguration either from a file or the default config.
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	"kythe.io/kythe/go/extractors/config/default/gradle"
+	"kythe.io/kythe/go/extractors/config/default/mvn"
+
+	"github.com/golang/protobuf/jsonpb"
+
+	ecpb "kythe.io/kythe/proto/extraction_config_go_proto"
+)
+
+func findConfig(configPath, repoDir string) (*ecpb.ExtractionConfiguration, error) {
+	// if a config was passed in, use the specified config, otherwise go
+	// hunt for one in the repository.
+	if configPath == "" {
+		// otherwise, use a Kythe config within the repo (if it exists)
+		configPath = filepath.Join(repoDir, kytheExtractionConfigFile)
+	}
+
+	f, err := os.Open(configPath)
+	if os.IsNotExist(err) {
+		// TODO(#3060): Depending on how we want to handle multiple types of
+		// builders in a single repo, this may be a bad approach.
+		// Specifically, we just greedily pick the first one we find and
+		// ignore others.
+		for _, b := range supportedBuilders() {
+			r, ok := testBuilder(b, repoDir)
+			if ok {
+				return load(r)
+			}
+		}
+		return nil, fmt.Errorf("failed to find a supported builder for repo %s", repoDir)
+	} else if err != nil {
+		return nil, fmt.Errorf("opening config file: %v", err)
+	}
+
+	defer f.Close()
+	return load(f)
+}
+
+func supportedBuilders() []builderType {
+	return []builderType{&mvn.Mvn{}, &gradle.Gradle{}}
+}
+
+// builderType gives info about a given builder.
+type builderType interface {
```

Comment:
I'd be inclined to make it a concrete type, unless you already need multiple implementations. It doesn't seem likely that the name needs to be computed, certainly, and maybe not the build file path either.

Perhaps:

```
type Builder struct {
   Name string
   ConfigFile string
   DefaultConfig func() (*kpb.ExtractionConfiguration, error)
}
```

The idea being here that if the user wants to override the default config, they can do whatever they want to construct one, and that's easier for testing than ginning up a reader. Also, it's worth letting it report an error for cases where something isn't found, isn't constructible, isn't readable, is invalid, etc. And then if this is `nil` the default can be some systemwide thing, or perhaps just an empty.

But then I wonder if it wouldn't be simpler to not distinguish between "config" and "default config" at all: Maybe it's up to the implementation whether there should be a fallback? In that case, an interface might be OK

```
type Builder interface {
  Name() string
  Config() (*kpb.ExtractionConfiguration, error)
}
```

I don't have a strong sense of the right answer here, but I'm biased toward as simple an interface as we can get away with.

Reply:
Separately: No need for `Type` on the end.

Reply:
Yea that first block is what I was thinkni gof.

We need to distinguish between normal build file name and DefaultConfig, because they're actually separate things.  The build file is e.g. 'build.gradle', but DefaultConfig is a fully specified ExtractionConfiguration proto, which contains a pointer to build.gradle but also a whole host of other things.

We need the build.gradle bit to be able to inspect the repo and see if it optimistically matches.

But I can go ahead and make the struct change real quick.

Reply:
or maybe 

```
type Builder struct {
  Name string
  DefaultConfig func(repo string) (*kpb.ExtractionConfiguration, error)
}
```

I dunno yet.

Reply:
Ok this is mostly reworked.  api should be cleaner now

also more tests